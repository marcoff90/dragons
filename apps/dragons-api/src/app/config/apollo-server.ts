import { ApolloServer } from 'apollo-server-express';
import { dragonSchema } from '../schemas/dragon-schema';
import { classSchema } from '../schemas/class-schema';
import { speciesSchema } from '../schemas/species-schema';
import { teamSchema } from '../schemas/team-image-schema';
import { dragonsResolver } from '../resolvers/dragons-resolver';
import { teamsResolver } from '../resolvers/teams-resolver';
import { classesResolver } from '../resolvers/classes-resolver';
import { speciesResolver } from '../resolvers/species-resolver';
import logger from '@dragons/util-logger';
import { querySchema } from '../schemas/query-schema';
import ClassRepository from '../repositories/class-repository';
import DragonRepository from '../repositories/dragon-repository';
import SpeciesRepository from '../repositories/species-repository';
import TeamImageRepository from '../repositories/team-image-repository';
import { buildSubgraphSchema } from '@apollo/subgraph';

export const server = new ApolloServer({
  schema: buildSubgraphSchema([
    {
      typeDefs: dragonSchema,
      resolvers: dragonsResolver,
    },
    {
      typeDefs: classSchema,
      resolvers: classesResolver,
    },
    {
      typeDefs: speciesSchema,
      resolvers: speciesResolver,
    },
    {
      typeDefs: teamSchema,
      resolvers: teamsResolver,
    },
    {
      typeDefs: querySchema,
    },
  ]),
  formatError: (error) => {
    logger.error(error);
    return new Error(error.message);
  },
  context: () => ({
    DragonRepository,
    ClassRepository,
    SpeciesRepository,
    TeamImageRepository,
  }),
});
