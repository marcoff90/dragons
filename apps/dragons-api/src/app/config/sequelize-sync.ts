import ClassModel from '../models/class-model';
import DragonModel from '../models/dragon-model';
import SpeciesModel from '../models/species-model';
import sequelize from './sequelize-config';

export const syncSequelize = async () => {
  ClassModel.hasMany(DragonModel, {
    as: 'dragons',
    foreignKey: 'class_id',
  });
  SpeciesModel.hasMany(DragonModel, {
    foreignKey: 'species_id',
  });
  DragonModel.belongsTo(ClassModel);
  DragonModel.belongsTo(SpeciesModel);
  await sequelize.sync();
};
