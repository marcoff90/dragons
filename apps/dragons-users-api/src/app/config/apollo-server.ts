import { buildSubgraphSchema } from '@apollo/subgraph';
import { ApolloServer } from 'apollo-server-express';
import { userSchema } from '../schemas/user-schema';
import { mutationSchema } from '../schemas/mutation-schema';
import logger from '@dragons/util-logger';
import UserService from '../services/user-service';
import { userResolver } from '../resolvers/user-resolver';

export const server = new ApolloServer({
  schema: buildSubgraphSchema([
    {
      typeDefs: userSchema,
      resolvers: userResolver,
    },
    {
      typeDefs: mutationSchema,
    },
  ]),
  formatError: (error) => {
    logger.error(error);
    return new Error(error.message);
  },
  context: () => ({
    UserService,
  }),
});
