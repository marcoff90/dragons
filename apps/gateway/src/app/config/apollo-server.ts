import {
  ApolloGateway,
  IntrospectAndCompose,
  RemoteGraphQLDataSource,
} from '@apollo/gateway';
import { ApolloServer } from 'apollo-server-express';
import logger from '@dragons/util-logger';

const gateway = new ApolloGateway({
  supergraphSdl: new IntrospectAndCompose({
    subgraphs: [
      { name: 'dragons', url: 'http://localhost:3333/graphql' },
      { name: 'users', url: 'http://localhost:4002/graphql' },
      { name: 'orders', url: 'http://localhost:4003/graphql' },
    ],
  }),
  buildService({ url }) {
    return new RemoteGraphQLDataSource({
      url,
      willSendRequest({ request, context }) {
        request.http.headers.set(
          'user',
          context.user ? JSON.stringify(context.user) : null
        );
      },
    });
  },
});

export const server = new ApolloServer({
  gateway,
  formatError: (error) => {
    logger.error(error);
    return new Error(error.message);
  },
  context: ({ req }) => {
    const user = req['user'] || null;
    return { user };
  },
});
