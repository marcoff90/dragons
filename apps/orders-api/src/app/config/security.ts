import { rule, shield } from 'graphql-shield';

const isAuthenticated = rule()((parent, args, { user }) => {
  return user !== null;
});

export const security = shield({
  Query: {
    orders: isAuthenticated,
    order: isAuthenticated,
  },
  Mutation: {
    createOrder: isAuthenticated,
  },
});
