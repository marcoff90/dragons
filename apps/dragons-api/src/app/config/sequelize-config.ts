import { Dialect, Sequelize } from 'sequelize';
import 'dotenv/config';

const { RDS_DB_NAME, RDS_USERNAME, RDS_PASSWORD, RDS_HOSTNAME, DIALECT } =
  process.env;

const sequelize: Sequelize = new Sequelize(
  RDS_DB_NAME,
  RDS_USERNAME,
  RDS_PASSWORD,
  {
    dialect: (DIALECT as Dialect) || 'postgres',
    host: RDS_HOSTNAME,
  }
);

export default sequelize;
