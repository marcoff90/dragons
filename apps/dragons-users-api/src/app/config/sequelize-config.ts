import { Dialect, Sequelize } from 'sequelize';
import 'dotenv/config';

const { RDS_DB_NAME_USERS, RDS_USERNAME, RDS_PASSWORD, RDS_HOSTNAME, DIALECT } =
  process.env;

const sequelize: Sequelize = new Sequelize(
  RDS_DB_NAME_USERS,
  RDS_USERNAME,
  RDS_PASSWORD,
  {
    dialect: (DIALECT as Dialect) || 'postgres',
    host: RDS_HOSTNAME,
  }
);

export default sequelize;
