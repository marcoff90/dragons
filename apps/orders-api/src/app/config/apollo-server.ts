import { ApolloServer } from 'apollo-server-express';
import { buildSubgraphSchema } from '@apollo/subgraph';
import OrderService from '../services/order-service';
import { mutationSchema } from '../schemas/mutation-schema';
import { orderSchema } from '../schemas/order-schema';
import { orderResolver } from '../resolvers/order-resolver';
import { orderItemSchema } from '../schemas/order-item-schema';
import { orderItemResolver } from '../resolvers/order-item-resolver';
import { querySchema } from '../schemas/query-schema';
import { applyMiddleware } from 'graphql-middleware';
import { security } from './security';
import logger from '@dragons/util-logger';

export const server = new ApolloServer({
  schema: applyMiddleware(
    buildSubgraphSchema([
      {
        typeDefs: orderItemSchema,
        resolvers: orderItemResolver,
      },
      {
        typeDefs: orderSchema,
        resolvers: orderResolver,
      },
      {
        typeDefs: mutationSchema,
      },
      {
        typeDefs: querySchema,
      },
    ]),
    security
  ),
  formatError: (error) => {
    logger.error(error);
    return new Error(error.message);
  },
  context: ({ req }) => {
    const user = req.headers.user
      ? JSON.parse(req.headers.user as string)
      : null;
    return { OrderService, user };
  },
});
