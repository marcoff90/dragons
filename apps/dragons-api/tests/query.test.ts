import { gql } from 'apollo-server-express';
import { createTestClient } from 'apollo-server-testing';
import { server } from '../src/app/config/apollo-server';

describe('Query tests', () => {
  const testClient = createTestClient(server);

  test('Dragons query no params', async () => {
    const DragonsQuery = gql(`
      {
        dragons {
          id
          name
        }
      }
    `);
    const { query } = testClient;
    const res = await query({ query: DragonsQuery });
    expect(res).toMatchSnapshot();
  });

  test('Dragons query with class and species', async () => {
    const DragonsQuery = gql(`
      {
        dragons {
          id
          name
          class
          species
        }
      }
    `);
    const { query } = testClient;
    const res = await query({ query: DragonsQuery });
    expect(res).toMatchSnapshot();
  });

  test('Dragons query with params', async () => {
    const DragonsQuery = gql(`
      {
        dragons(limits: { limit: 5 }) {
          id
          name
        }
      }
    `);
    const { query } = testClient;
    const res = await query({ query: DragonsQuery });
    expect(res).toMatchSnapshot();
  });

  test('Dragon query with name param', async () => {
    const DragonQuery = gql(`
      {
        dragon(input: { name: "Toothless" }) {
          id
        }
      }
    `);
    const { query } = testClient;
    const res = await query({ query: DragonQuery });
    expect(res).toMatchSnapshot();
  });

  test('Dragon query with no param returns Error', async () => {
    const DragonQuery = gql(`
      {
        dragon {
          id
        }
      }
    `);
    const { query } = testClient;
    const res = await query({ query: DragonQuery });
    expect(res).toMatchSnapshot();
    expect(res.errors[0].message).toEqual(
      'Field "dragon" argument "input" of type "OneDragon!" is required, but it was not provided.'
    );
  });

  test('Dragon query with no no name in db returns Error', async () => {
    const DragonQuery = gql(`
      {
        dragon(input: { name: "HelloWorld" }) {
          id
        }
      }
    `);
    const { query } = testClient;
    const res = await query({ query: DragonQuery });
    expect(res).toMatchSnapshot();
    expect(res.errors[0].message).toEqual('HelloWorld not found');
  });

  test('Distinct heads returns values for filters', async () => {
    const HeadQuery = gql(`
      {
        distinctHeads {
          heads
        }
      }
    `);
    const { query } = testClient;
    const res = await query({ query: HeadQuery });
    expect(res).toMatchSnapshot();
  });

  test('Teams query', async () => {
    const TeamsQuery = gql(`
      {
        teams {
          id
          image
        }
      }
    `);
    const { query } = testClient;
    const res = await query({ query: TeamsQuery });
    expect(res).toMatchSnapshot();
  });

  test('Classes query', async () => {
    const ClassesQuery = gql(`
      {
        classes {
          id
          name
          image
        }
      }
    `);
    const { query } = testClient;
    const res = await query({ query: ClassesQuery });
    expect(res).toMatchSnapshot();
  });

  test('Species query', async () => {
    const SpeciesQuery = gql(`
      {
        species {
          id
          species
        }
      }
    `);
    const { query } = testClient;
    const res = await query({ query: SpeciesQuery });
    expect(res).toMatchSnapshot();
  });
});
