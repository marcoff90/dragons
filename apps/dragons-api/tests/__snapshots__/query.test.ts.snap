// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Query tests Classes query 1`] = `
Object {
  "data": Object {
    "classes": Array [
      Object {
        "id": 1,
        "image": "http://localhost:3333/assets/dragon-classes/strike.png",
        "name": "Strike",
      },
      Object {
        "id": 2,
        "image": "http://localhost:3333/assets/dragon-classes/boulder.png",
        "name": "Boulder",
      },
      Object {
        "id": 3,
        "image": "http://localhost:3333/assets/dragon-classes/mystery.png",
        "name": "Mystery",
      },
      Object {
        "id": 4,
        "image": "http://localhost:3333/assets/dragon-classes/sharp.jpg",
        "name": "Sharp",
      },
      Object {
        "id": 5,
        "image": "http://localhost:3333/assets/dragon-classes/tidal.png",
        "name": "Tidal",
      },
      Object {
        "id": 6,
        "image": "http://localhost:3333/assets/dragon-classes/tracker.png",
        "name": "Tracker",
      },
      Object {
        "id": 7,
        "image": "http://localhost:3333/assets/dragon-classes/stoker.png",
        "name": "Stoker",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Query tests Distinct heads returns values for filters 1`] = `
Object {
  "data": Object {
    "distinctHeads": Object {
      "heads": Array [
        4,
        1,
        2,
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Query tests Dragon query with name param 1`] = `
Object {
  "data": Object {
    "dragon": Object {
      "id": 1,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Query tests Dragon query with no no name in db returns Error 1`] = `
Object {
  "data": Object {
    "dragon": null,
  },
  "errors": Array [
    Object {
      "message": "HelloWorld not found",
    },
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Query tests Dragon query with no param returns Error 1`] = `
Object {
  "data": undefined,
  "errors": Array [
    Object {
      "message": "Field \\"dragon\\" argument \\"input\\" of type \\"OneDragon!\\" is required, but it was not provided.",
    },
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Query tests Dragons query no params 1`] = `
Object {
  "data": Object {
    "dragons": Array [
      Object {
        "id": 1,
        "name": "Toothless",
      },
      Object {
        "id": 2,
        "name": "Stormfly",
      },
      Object {
        "id": 3,
        "name": "Barf & Belch",
      },
      Object {
        "id": 4,
        "name": "Grump",
      },
      Object {
        "id": 5,
        "name": "HookFang",
      },
      Object {
        "id": 6,
        "name": "Light Fury",
      },
      Object {
        "id": 7,
        "name": "Meatlug",
      },
      Object {
        "id": 8,
        "name": "CloudJumper",
      },
      Object {
        "id": 9,
        "name": "SkullCrusher",
      },
      Object {
        "id": 10,
        "name": "ArmorWing",
      },
      Object {
        "id": 11,
        "name": "Eruptodon",
      },
      Object {
        "id": 12,
        "name": "Night Terror",
      },
      Object {
        "id": 13,
        "name": "SeaShocker",
      },
      Object {
        "id": 14,
        "name": "Skrill",
      },
      Object {
        "id": 15,
        "name": "SnapTrapper",
      },
      Object {
        "id": 16,
        "name": "Terrible Terror",
      },
      Object {
        "id": 17,
        "name": "ThunderDrum",
      },
      Object {
        "id": 18,
        "name": "TimberJack",
      },
      Object {
        "id": 19,
        "name": "Whispering Death",
      },
      Object {
        "id": 20,
        "name": "Baby Gronckle",
      },
      Object {
        "id": 21,
        "name": "Baby Nadder",
      },
      Object {
        "id": 22,
        "name": "Baby Nightmare",
      },
      Object {
        "id": 23,
        "name": "Baby Zippleback",
      },
      Object {
        "id": 24,
        "name": "Drago's BeWilderBeast",
      },
      Object {
        "id": 25,
        "name": "Valka's BeWilderBeast",
      },
      Object {
        "id": 26,
        "name": "Death Song",
      },
      Object {
        "id": 27,
        "name": "Dramillion",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Query tests Dragons query with class and species 1`] = `
Object {
  "data": undefined,
  "errors": Array [
    Object {
      "message": "Field \\"class\\" of type \\"Class!\\" must have a selection of subfields. Did you mean \\"class { ... }\\"?",
    },
    Object {
      "message": "Field \\"species\\" of type \\"Species!\\" must have a selection of subfields. Did you mean \\"species { ... }\\"?",
    },
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Query tests Dragons query with params 1`] = `
Object {
  "data": Object {
    "dragons": Array [
      Object {
        "id": 1,
        "name": "Toothless",
      },
      Object {
        "id": 2,
        "name": "Stormfly",
      },
      Object {
        "id": 3,
        "name": "Barf & Belch",
      },
      Object {
        "id": 4,
        "name": "Grump",
      },
      Object {
        "id": 5,
        "name": "HookFang",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Query tests Species query 1`] = `
Object {
  "data": Object {
    "species": Array [
      Object {
        "id": 1,
        "species": "Night Fury",
      },
      Object {
        "id": 2,
        "species": "Deadly Nadder",
      },
      Object {
        "id": 3,
        "species": "Hideous Zippleback",
      },
      Object {
        "id": 4,
        "species": "HotBurple",
      },
      Object {
        "id": 5,
        "species": "Monstrous nightmare",
      },
      Object {
        "id": 6,
        "species": "Light Fury",
      },
      Object {
        "id": 7,
        "species": "Gronckle",
      },
      Object {
        "id": 8,
        "species": "StormCuter",
      },
      Object {
        "id": 9,
        "species": "RumbleHorn",
      },
      Object {
        "id": 10,
        "species": "ArmorWing",
      },
      Object {
        "id": 11,
        "species": "Eruptodon",
      },
      Object {
        "id": 12,
        "species": "Night Terror",
      },
      Object {
        "id": 13,
        "species": "SeaShocker Terror",
      },
      Object {
        "id": 14,
        "species": "Skrill",
      },
      Object {
        "id": 15,
        "species": "SnapTrapper",
      },
      Object {
        "id": 16,
        "species": "Terrible Terror",
      },
      Object {
        "id": 17,
        "species": "ThunderDrum",
      },
      Object {
        "id": 18,
        "species": "TimberJack",
      },
      Object {
        "id": 19,
        "species": "Whispering Death",
      },
      Object {
        "id": 20,
        "species": "BeWilderBeast",
      },
      Object {
        "id": 21,
        "species": "Death Song",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Query tests Teams query 1`] = `
Object {
  "data": Object {
    "teams": Array [
      Object {
        "id": 1,
        "image": "http://localhost:3333/assets/teams/barf_and_belch_dragon-team.png",
      },
      Object {
        "id": 2,
        "image": "http://localhost:3333/assets/teams/hookfang_dragon-team.png",
      },
      Object {
        "id": 3,
        "image": "http://localhost:3333/assets/teams/meatlug_dragon-team.png",
      },
      Object {
        "id": 4,
        "image": "http://localhost:3333/assets/teams/stormfly_dragon-team",
      },
      Object {
        "id": 5,
        "image": "http://localhost:3333/assets/teams/toothles_dragon-team.png",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
